### PR TITLE
Feature/ta new protocol need

### PR DESCRIPTION
**Pull request by timigod**
## Description
- Renamed `RequestController` to `NeedController`
- Changed associated `Need` related routes (also stopped using GET for everything)
- Refractored stores and everything else to use `need` instead of `request`
- Refractored stores and everything else to use new parameters specified in the documentation.
- Added exhaustive validation to need creation endpoint.

### CODE REVIEW

File: server/controllers/NeedController.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
+const {createNeed, getNeed, deleteNeed} = require('../store/needs');
+const {deleteBidsForNeed} = require('../store/bids');
+const {createMission} = require('../store/missions');
+const {updateVehicleStatus} = require('../store/vehicles');
+const createConstraints = require('./constraints/need/create');
+
+// using downloaded validate.js file because latest version does not have type checking and docker fails with git source in package.json
```
Comment (TalAter) [inline server/controllers/NeedController.js]: You can delete this comment.
It's interesting during the review time, and less so for posterity.
If you'd like future generations to understand your reasoning, maybe include it on the top of the validation file itself.

File: server/controllers/constraints/need/create.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
+module.exports = {
+  pickup_at: {
+    numericality: {
+      greaterThan: Date.now()
```
Comment (TalAter) [inline server/controllers/constraints/need/create.js]: or equal to

File: server/controllers/constraints/need/create.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    }
+  },
+  pickup_latitude: {
+    presence: true,
+    numericality: {
+      lessThanOrEqualTo: 90,
```
Comment (TalAter) [inline server/controllers/constraints/need/create.js]: can we also add a minimum to the range?

File: server/controllers/constraints/need/create.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    }
+  },
+  pickup_longitude: {
+    presence: true,
+    numericality: {
+      lessThanOrEqualTo: 180,
```
Comment (TalAter) [inline server/controllers/constraints/need/create.js]: can we also add a minimum to the range?

File: server/controllers/constraints/need/create.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    }
+  },
+  dropoff_latitude: {
+    presence: true,
+    numericality: {
+      lessThanOrEqualTo: 90,
```
Comment (TalAter) [inline server/controllers/constraints/need/create.js]: can we also add a minimum to the range?

File: server/controllers/constraints/need/create.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    }
+  },
+  dropoff_longitude: {
+    presence: true,
+    numericality: {
+      lessThanOrEqualTo: 180,
```
Comment (TalAter) [inline server/controllers/constraints/need/create.js]: can we also add a minimum to the range?

File: server/controllers/constraints/need/create.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    }
+  },
+  requester_phone_number: {
+    length: {
+      minimum: 8,
+    }
+  },
+  cargo_type: {
+    presence: true,
+    type: 'number',
```
Comment (TalAter) [inline server/controllers/constraints/need/create.js]: will it work with a numeric string?
Same question for next ones. All numeric values may come as strings

File: server/controllers/constraints/need/create.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    inclusion: {
```
Comment (TalAter) [inline server/controllers/constraints/need/create.js]: Wouldn't it be easier to just say it needs to be between 1 and 18?
Same question for next ones

File: server/controllers/constraints/need/create.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    }
+  },
+  ip_protection_level: {
+    type: 'number',
+    inclusion: {
+      within: Array.from(new Array(69), (x, i) => i + 54)
+    }
+  },
+  height: {
+    type: 'number'
```
Comment (TalAter) [inline server/controllers/constraints/need/create.js]: greater than 0
(same for other measurements)

File: server/controllers/constraints/need/create.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  length: {
+    type: 'number'
+  },
+  weight: {
+    type: 'number'
+  },
+  insurance_required: {
+    type: 'boolean'
+  },
+  insured_value: {
```
Comment (TalAter) [inline server/controllers/constraints/need/create.js]: greater than 0

File: server/store/missions.js [hunk 11]
```diff
@@ -51,13 +51,13 @@ const createMission = async ({ user_id, bid_id }) => {
...
     'time_to_pickup', time_to_pickup,
     'time_to_dropoff', time_to_dropoff,
-    'request_id', request_id,
-    'pickup_lat', pickup_lat,
-    'pickup_long', pickup_long,
-    'dropoff_lat', dropoff_lat,
-    'dropoff_long', dropoff_long,
-    'requested_pickup_time', requested_pickup_time,
-    'size', size,
+    'need_id', need_id,
```
Comment (TalAter) [inline server/store/missions.js]: We are updating the communication docs from `request_uid` to `need_id` as you suggest here 👍

File: server/controllers/NeedController.js (hunk 1 above)
Comment (timigod) [reply server/controllers/NeedController.js]: Will include it in the validation file.

File: server/controllers/constraints/need/create.js (hunk 3 above)
Comment (timigod) [reply server/controllers/constraints/need/create.js]: Yes we can, -90 right?

File: server/controllers/constraints/need/create.js (hunk 8 above)
Comment (timigod) [reply server/controllers/constraints/need/create.js]: It would. Thanks!

File: server/store/missions.js (hunk 11 above)
Comment (timigod) [reply server/store/missions.js]: Great!

File: server/controllers/constraints/need/create.js (hunk 7 above)
Comment (timigod) [reply server/controllers/constraints/need/create.js]: It doesn't work for numeric strings. Not sure how to solve that.

Comment (timigod) [reply server/controllers/constraints/need/create.js]: It does work for numeric strings if I used the `numericality` validation instead of type.

**Pull request merged**